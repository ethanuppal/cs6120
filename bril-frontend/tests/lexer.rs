use bril_frontend::{lexer::Token, loc::Loc};
use insta::assert_debug_snapshot;
use logos::Logos;

#[allow(dead_code)] // since Debug
#[derive(Debug)]
struct LexerSnapshot<'a> {
    code: &'a str,
    tokens: Vec<Loc<Token<'a>>>,
}

#[allow(dead_code)] // since Debug
#[derive(Debug)]
struct LexerErrorSnapshot<'a> {
    code: &'a str,
    tokens: Vec<Loc<Token<'a>>>,
    failure: Loc<&'a str>,
    leftover: &'a str,
}

macro_rules! lexer_snapshot {
    ($name:ident, $code:expr) => {
        #[test]
        fn $name() {
            let code = $code;
            let mut lexer = Token::lexer(code);
            let mut tokens = vec![];
            while let Some(next) = lexer.next() {
                if let Ok(token) = next {
                    tokens.push(Loc::new(token, lexer.span()));
                } else {
                    panic!("Failed to lex. Leftover: {}", lexer.remainder());
                }
            }

            assert_debug_snapshot!(LexerSnapshot { code, tokens });
        }
    };
}

macro_rules! lexer_error {
    ($name:ident, $code:expr) => {
        #[test]
        fn $name() {
            let code = $code;
            let mut lexer = Token::lexer(code);
            let mut tokens = vec![];
            while let Some(next) = lexer.next() {
                if let Ok(token) = next {
                    tokens.push(Loc::new(token, lexer.span()));
                } else {
                    break;
                }
            }

            if lexer.remainder().is_empty() {
                panic!("Lexing was unfortunately successful: {:?}", tokens);
            }

            assert_debug_snapshot!(LexerErrorSnapshot {
                code,
                tokens,
                failure: Loc::new(lexer.slice(), lexer.span()),
                leftover: lexer.remainder()
            });
        }
    };
}

lexer_snapshot! {
    all_tokens_work,
    r#"import from as @main .foo foo "path/to/lol" {}:<>;=,() 5 5.0 'a'"#
}

lexer_snapshot! {
    add_bril_lexes,
    include_str!("../bril-programs/add.bril")
}

lexer_snapshot! {
    import_bril_lexes,
    include_str!("../bril-programs/import.bril")
}

lexer_error! {
    invalid_characters,
    "$main"
}
