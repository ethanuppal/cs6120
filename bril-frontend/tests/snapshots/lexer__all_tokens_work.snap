---
source: bril-frontend/tests/lexer.rs
expression: "LexerSnapshot { code, tokens }"
---
LexerSnapshot {
    code: "import from as @main .foo foo \"path/to/lol\" {}:<>;=,() 5 5.0 'a'",
    tokens: [
        Loc {
            inner: Import,
            span: 0..6,
        },
        Loc {
            inner: From,
            span: 7..11,
        },
        Loc {
            inner: As,
            span: 12..14,
        },
        Loc {
            inner: FunctionName(
                "@main",
            ),
            span: 15..20,
        },
        Loc {
            inner: Label(
                ".foo",
            ),
            span: 21..25,
        },
        Loc {
            inner: Identifier(
                "foo",
            ),
            span: 26..29,
        },
        Loc {
            inner: Path(
                "path/to/lol",
            ),
            span: 30..43,
        },
        Loc {
            inner: LeftBrace,
            span: 44..45,
        },
        Loc {
            inner: RightBrace,
            span: 45..46,
        },
        Loc {
            inner: Colon,
            span: 46..47,
        },
        Loc {
            inner: LeftAngle,
            span: 47..48,
        },
        Loc {
            inner: RightAngle,
            span: 48..49,
        },
        Loc {
            inner: Semi,
            span: 49..50,
        },
        Loc {
            inner: Equals,
            span: 50..51,
        },
        Loc {
            inner: Comma,
            span: 51..52,
        },
        Loc {
            inner: LeftPar,
            span: 52..53,
        },
        Loc {
            inner: RightPar,
            span: 53..54,
        },
        Loc {
            inner: Integer(
                5,
            ),
            span: 55..56,
        },
        Loc {
            inner: Float(
                5.0,
            ),
            span: 57..60,
        },
        Loc {
            inner: Character(
                'a',
            ),
            span: 61..64,
        },
    ],
}
